### ISSUE TITLE
Project dependencies may have API risk issues

### ISSUE THREAD

**Issue by PyDeps** [Tomiinek/Multilingual_Text_to_Speech#82]
Hi, In **Multilingual_Text_to_Speech**, inappropriate dependency versioning constraints can cause risks.

Below are the dependencies and version constraints that the project is using

```
jupyter==1.0.0
editdistance==0.5.3
epitran==1.2
fastdtw==0.3.4
future
jieba==0.42.1
librosa==0.7.0
matplotlib
numpy
pandas
phonemizer==2.1
Pillow==9.0.1
pinyin==0.4.0
scikit-learn
scipy
seaborn
SoundFile==0.10.2
tensorboard==2.0.2
torch==1.3.0
numba==0.48
```

The version constraint **==** will introduce the risk of dependency conflicts because the scope of dependencies is too strict.
The version constraint **No Upper Bound** and **\*** will introduce the risk of the missing API Error because the latest version of the dependencies may remove some APIs.

After further analysis, in this project,
The version constraint of dependency **fastdtw** can be changed to *>=0.1.0,<=0.2.2*.
The version constraint of dependency **future** can be changed to *>=0.12.0,<=0.18.2*.
The version constraint of dependency **jieba** can be changed to *>=0.36,<=0.36.2*.
The version constraint of dependency **scipy** can be changed to *>=0.12.0,<=1.7.3*.
The version constraint of dependency **SoundFile** can be changed to *>=0.8.1,<=0.9.0.post1*.
The version constraint of dependency **numba** can be changed to *>=0.7.2,<=0.11.0*.

The above modification suggestions can reduce the dependency conflicts as much as possible, 
and introduce the latest version as much as possible without calling Error in the projects.

The invocation of the current project includes all the following methods.

<details><summary>The calling methods from the fastdtw</summary>
 <pre>fastdtw.fastdtw
</pre>
</details>
<details><summary>The calling methods from the future</summary>
 <pre>datetime.datetime.now
</pre>
</details>
<details><summary>The calling methods from the jieba</summary>
 <pre>jieba.cut
</pre>
</details>
<details><summary>The calling methods from the scipy</summary>
 <pre>scipy.stats.sem
scipy.signal.lfilter
</pre>
</details>
<details><summary>The calling methods from the SoundFile</summary>
 <pre>soundfile.write
soundfile.read
</pre>
</details>
<details><summary>The calling methods from the numba</summary>
 <pre>random.seed
</pre>
</details>
<details><summary>The calling methods from the all methods</summary>
 <pre>self._embedding
torch.no_grad
numpy.average
self._lstm.flatten_parameters
Logger._plot_alignment
self._target_init.transpose
utils.logging.Logger.evaluation
torch.nn.functional.conv1d
datetime.datetime.now.strftime
self._kernel
layer
modules.tacotron2.Tacotron.eval
state_dict.items
set
matplotlib.pyplot.show
idx.target_len.idx.prediction.data.cpu.numpy
glob.glob
name.lower
db_to_amplitude
self._convolution
modules.layers.ConvBlockGenerated
self._get_attention
idx.target_len.idx.prediction.data.cpu
google.protobuf.json_format.MessageToDict
model
params.params.Params.languages.index
self.ConvBlock.super.__init__
IOError
librosa.feature.inverse.mel_to_stft
torch.div
l.s.t.model.inference.cpu.detach
modules.encoder.ConditionalEncoder
torch.argmax
stop_idx.gen.data.cpu.numpy
set.add
self.ZoneoutLSTMCell.super.__init__
params.params.Params.symbols_count
self.LocationSensitiveAttention.super.__init__
self._num_features.affine.contiguous
torch.sort
os.path.splitext
numpy.load
self._attention_lstm
print
torch.argmax.unsqueeze
os.path.dirname
torch.nn.functional.mse_loss
torch.nn.ModuleList
list
modules.generated.Conv1dGenerated
Decoder
self._prenet
a.split
self._get_adversarial_classifier
torch.tensor
l.s.t.model.inference.cpu
enumerate
matplotlib.pyplot.figure.add_subplot
torch.optim.lr_scheduler.StepLR.load_state_dict
torch.optim.Adam.zero_grad
str
idx.source.data.cpu
self.TacotronLoss.super.__init__
Params.state_dict
params.params.Params.load
ast.literal_eval
stop_token.masked_fill_
mcds.append
train
torch.nn.Tanh
get_spectrogram_mfcc
items.append
speakers.unsqueeze.expand.unsqueeze
self._activation
torch.nn.Sigmoid
grad_output.clamp.clamp
self._highway_layers
i.str.zfill
utils.lengths_to_mask
params.params.Params.state_dict
utils.text.remove_punctuation
torch.zeros.squeeze
torch.nn.ConstantPad1d
x.transpose.reshape
line.rstrip
self._output_layer
x.transpose.size
input_data.split.split
ml.speakers.repeat.transpose
romkan.to_roma
self._generator_lstm
idx.stop_prediction.data.cpu.numpy
fig.add_subplot.imshow
self.ForwardAttentionWithTransition.super.__init__
super.__getattr__
speakers.unsqueeze.expand.dim
clean
utils.text.to_phoneme
modules.attention.ForwardAttentionWithTransition
os.path.isfile
self.HighwayConvBlockGenerated.super.__init__
cum_weights.unsqueeze.transpose
zip
matplotlib.pyplot.figure.colorbar
self._linear
torch.LongTensor.add
getattr
torch.sigmoid
Logger._sw.add_scalar
cum_weights.unsqueeze.unsqueeze
self.items.append
numpy.array
text.unsqueeze_
scipy.signal.lfilter
google.cloud.speech_v1.SpeechClient.recognize
d.items
torch.exp
sys.path.insert
tqdm.tqdm.set_postfix_str
torch.nn.Parameter
google.cloud.speech_v1.SpeechClient
utils.audio.mel_cepstral_distorision
self.register_buffer
torch.manual_seed
float
librosa.feature.mfcc
numpy.where
torch.nn.functional.mse_loss.values
torch.FloatTensor
torch.optim.Adam
self.Encoder.super.__init__
utils.audio.save
self.ZoneoutLSTMCell.super.forward
librosa.griffinlim
datetime.datetime.now
self._prepare_transition
Logger._plot_stop_tokens
self._frame_prediction
dataset.dataset.TextToSpeechCollate
modules.tacotron2.Tacotron.state_dict
l.cuda.rstrip
torch.arange
os.path.join
idx.source_len.idx.target_len.idx.alignment.data.cpu
self._add_conditional_embedding
os.path.exists
params.params.Params.unique_speakers.index
min
modules.tacotron2.Tacotron.train
speaker.unsqueeze.expand.unsqueeze
numpy.min
self._transition_agent
prediction.transpose.transpose
librosa.amplitude_to_db
one_hot.scatter_.append
l.rstrip
collections.OrderedDict
dataset.train.get_num_speakers
utils.audio.normalize_spectrogram
levenshtein
next
modules.tacotron2.TacotronLoss.state_dict
torch.nn.functional.dropout
self._energy
t.remove_punctuation.split
self._location.transpose
librosa.get_duration
self._bias
ReversalClassifier.loss
model.state_dict.update
torch.tanh
epi.transliterate.replace
self.ConditionalEncoder.super.__init__
modules.tacotron2.Tacotron.cuda
self._affine
utils.text.to_lower
utils.logging.Logger.training
e.self._kernel.view
e.self._bias.view
loss.backward
torch.max
modules.classifier.ReversalClassifier.loss
Logger._plot_spectrogram
self._get_embedding
model._prenet.parameters
random.randint
modules.layers.HighwayConvBlock
folder.replace
matplotlib.pyplot.tight_layout
librosa.feature.melspectrogram
self.ForwardAttentionWithTransition.super.forward
losses.items
dir
modules.generated.BatchNorm1dGenerated
modules.attention.ForwardAttention
utils.text.to_text
GradientReversalFunction.apply
epitran.Epitran
torch.sum
text.split
j.trg_len.ref.data.cpu
predictions.data.cpu.numpy
self._attention.reset
self._bottleneck
deemphasis
_phonemize
self.ForwardAttentionWithTransition.super.reset
synthesize
torch.nn.Linear
one_hot.scatter_.size
pinyin.get
torch.optim.Adam.load_state_dict
ttf.read
batch_losses.items
sample_cer
self._gate
phonemizer.phonemize.phonemize
utils.text.remove_odd_whitespaces
torch.nn.functional.pad
torch.ones
torch.load.items
known_unique_speakers.copy
l_cw.split
modules.layers.ZoneoutLSTMCell
self.Prenet.super.__init__
self._layer_pass
meta_file.os.path.basename.split
sum
self.MultiEncoder.super.__init__
utils.build_model.eval
dataset.loaders.get_loader_by_name
self._regularizer
modules.classifier.CosineSimilarityClassifier
modules.tacotron2.Tacotron
f.read
self._attent
sys.stdin.readlines
predictions.data.cpu
self._decoder.inference
epi.transliterate.append
df.read
fileinput.input
self._query.unsqueeze
self.PostnetCBHG.super.__init__
self._sampler.__iter__
self._num_features.affine.contiguous.view
torch.zeros_like
target_mask.unsqueeze.float
torch.norm
eval_losses.values
self.DropoutLSTMCell.super.__init__
self._bias.energy.torch.tanh.self._energy.squeeze
spectrogram
torch.optim.Adam.state_dict
self._block
punct_re.replace
inverse_spectrogram
embeddings.torch.norm.unsqueeze
params.params.Params.load_state_dict
get_activation
b.append
utils.audio.denormalize_spectrogram
utils.text.to_sequence
mel_cepstral_distorision
numpy.sum
torch.nn.functional.softmax
x.contiguous.cuda
dataset_name.loaders.get_loader_by_name
Postnet
torch.save
torch.nn.functional.binary_cross_entropy_with_logits
modules.layers.ConvBlock
numpy.std
numpy.sqrt
get_mfcc
torch.nn.GRU
self.ForwardAttention.super.__init__
torch.LongTensor
w.torch.norm.view
input_mask.torch.sum.item
attention_weights.unsqueeze.unsqueeze
eval_losses.keys
utils.samplers.RandomImbalancedSampler
torch.nn.utils.rnn.pad_packed_sequence
torch.nn.BatchNorm1d
idx.stop_prediction.data.cpu
token.split.split
self._decoder
TextToSpeechDataset
evaluate
tqdm.tqdm
modules.encoder.Encoder
self._guided_attention
torch.LongTensor.append
self._postnet.squeeze
self._convs
audio.mel_spectrogram
self.Postnet.super.__init__
input_data.split
utils.audio.inverse_spectrogram
torch.abs
self.BatchNorm1dGenerated.super.__init__
utils.samplers.PerfectBatchSampler
preemphasis
one_hot.scatter_
numpy.abs
json.load
self._location
modules.tacotron2.Tacotron.parameters
torch.rand
modules.tacotron2.Tacotron.to
librosa.stft
callable
self._encoder
self._add_conditional_embedding.size
idx.target_len.idx.target.data.cpu
modules.tacotron2.Tacotron.load_state_dict
join.split
stop_logits.torch.sigmoid.ge
modules.encoder.MultiEncoder
self.AttentionBase.super.__init__
x.contiguous.contiguous
memory.attention_weights.torch.bmm.squeeze
self._language_embedding
torch.chunk
self._classifier
lebel_freq.values
modules.tacotron2.TacotronLoss.update_states
numpy.shape
idx.stop_target.data.cpu.numpy
re.sub
self.ForwardAttention.super.reset
i.stop_tokens.squeeze
self._get_postnet
l.s.t.model.inference.cpu.detach.numpy
target.device.self._output_dim.batch_size.torch.zeros.unsqueeze
torch.argmax.sum
torch.clamp
losses.values
Logger._sw.add_audio
torch.nn.Embedding
modules.layers.HighwayConvBlockGenerated
torch.nn.LSTM
modules.classifier.CosineSimilarityClassifier.loss
Params.state_dict.items
torch.mean
torch.zeros
torch.load
utils.logging.Logger.progress
criterion
DataParallelPassthrough
modules.tacotron2.TacotronLoss.load_state_dict
text.lower
io.open
HighwayLayer
torch.cuda.device_count
model._reversal_classifier.parameters
idx.stop_target.data.cpu
grad_output.clamp.neg
SubsetSampler
remove_punctuation
self.ConvBlockGenerated.super.__init__
sample_recognize
fig.add_subplot.scatter
idx.source_len.idx.target_len.idx.alignment.data.cpu.numpy.size
l.rstrip.split
argparse.ArgumentParser
utils.text.build_phoneme_dicts
cers.append
ms_to_frames
self._pool_and_project
speaker.unsqueeze.expand
token.split
utils.audio.spectrogram
torch.optim.lr_scheduler.StepLR.state_dict
librosa.effects.trim
time.time
self.Decoder.super.__init__
self.ConvolutionalEncoder.super.__init__
map
l.x_langs_normed.reshape
line.split
GradientClippingFunction.apply
GradientClippingFunction.apply.view_as
synthesize.cuda
torch.utils.data.sampler.SubsetRandomSampler
w.bool.any
torch.utils.tensorboard.SummaryWriter
random.seed
matplotlib.pyplot.ylabel
self._target_init.size
confidence_interval
s.rstrip
modules.layers.DropoutLSTMCell
MeCab.Tagger
self._postnet
torch.nn.ReLU
idx.target_len.idx.prediction_forced.data.cpu.numpy
self.DropoutLSTMCell.super.forward
label.label_indices.append
self._stop_prediction
line.rstrip.split
matplotlib.pyplot.colorbar
self.CosineSimilarityClassifier.super.__init__
t.cuda.cuda
remove_dataparallel_prefix
speakers.unsqueeze.expand
os.makedirs
fastdtw.fastdtw
json.dump
torch.optim.lr_scheduler.StepLR.step
self.unique_speakers.index
utils.build_model.inference
attr.startswith
target_mask.unsqueeze.float.unsqueeze
model._attention.parameters
self._lstm
remove_punctuation.split
self.Tacotron.super.__init__
torch.nn.Dropout
torch.nn.functional.normalize
torch.device
self._layers
os.path.basename
matches.torch.sum.item
self._gru
join.lower
librosa.db_to_amplitude
join
epitran.Epitran.transliterate
self._normalize
glob.glob.sort
torch.bmm
metadata.append
modules.tacotron2.TacotronLoss
self._dropout
MeCab.Tagger.parse
model._postnet.parameters
torch.nn.Conv1d
soundfile.read
open
Params.load_state_dict
prediction.transpose.size
matplotlib.pyplot.title
idx.target_len.idx.prediction_forced.data.cpu
torch.nn.MaxPool1d
librosa.display.specshow
torch.cuda.is_available
utils.audio.load
l.x_langs_normed.reshape.bool
attention_weights.unsqueeze.squeeze
_other_symbols
prediction.transpose
self.Conv1dGenerated.super.__init__
x.transpose.transpose
soundfile.write
super
Logger._sw.add_text
self._query
self.HighwayConvBlock.super.__init__
languages.unsqueeze.expand
numpy.mean
x.transpose.expand
dataset.dataset.TextToSpeechDatasetCollection
sequence.append
torch.optim.Adam.step
numpy.save
int
one_hot.scatter_.add
matplotlib.pyplot.figure
j.trg_len.ref.data.cpu.numpy
len
torch.nn.Sequential
utils.audio.linear_to_mel.size
tuple
modules.encoder.ConvolutionalEncoder
utils.audio.linear_to_mel
stop_idx.gen.data.cpu
self._loc_features
torch.nn.Identity
max
self._reversal_classifier
torch.nn.utils.rnn.pack_padded_sequence
iter
Prenet
math.cos
modules.attention.LocationSensitiveAttention
self._decode
speaker.unsqueeze.expand.dim
amplitude_to_db
self._memory
scipy.stats.sem
idx.source_len.idx.target_len.idx.alignment.data.cpu.numpy
l.self._encoders
phonemizer.separator.Separator
modules.classifier.ReversalClassifier
self._gru.flatten_parameters
language.unsqueeze.expand
text.size
self.unique_speakers.append
torch.utils.data.sampler.WeightedRandomSampler
self.GeneratedConvolutionalEncoder.super.__init__
idx.target_len.idx.target.data.cpu.numpy
stop.cpu.numpy
torch.optim.lr_scheduler.StepLR
numpy.random.seed
self._padding
Logger._sw.add_figure
torch.nn.init.xavier_uniform_
jieba.cut
matplotlib.pyplot.xlabel
speakers.repeat
params.params.Params.language_number.languages.size.languages.size.torch.zeros.zero_
idx.source.data.cpu.numpy
self.HighwayConvBlock.super.forward
self.ReversalClassifier.super.__init__
torch.nn.functional.cross_entropy
self.HighwayConvBlockGenerated.super.forward
self._combine_weights
modules.encoder.GeneratedConvolutionalEncoder
model._encoder.parameters
dataset.train.get_normalization_constants
self._get_encoder
self.HighwayLayer.super.__init__
torch.nn.utils.clip_grad_norm_
setattr
modules.cbhg.PostnetCBHG
scipy.stats.t.ppf
torch.utils.data.DataLoader
model._embedding.parameters
self._attention
torch.nn.functional.batch_norm
argparse.ArgumentParser.parse_args
stop.cpu
meta_file.os.path.basename.split.split
cos_decay
torch.meshgrid
self.load_spectrogram
utils.build_model
torch.argmax.dim
argparse.ArgumentParser.add_argument
mcd
self._decoder_init
range
target_lengths.float
model._decoder.parameters
Encoder
self._target_init
torch.nn.functional.softmax.squeeze
utils.logging.Logger.initialize
l.cuda.cuda
torch.cat
</pre>
</details>

@developer
Could please help me check this issue?
May I pull a request to fix it?
Thank you very much.

**Comment by Tomiinek** [comment]
Hello, yeas please, feel free to introduce a PR :slightly_smiling_face:
